Add support for CodeDeploy Commands service

drake-amz opened the pull request:
See #1023 for more details.

This PR adds support for the CodeDeploy Commands service, used by the CodeDeploy agent to pull work. As this is not a service which is included in botocore, I've also added support to the `service_crategen` crate to load what I'm calling "embedded" service definitions. Embedded services have their `service-2.json` file embedded directly in the Rusoto repository, rather than being loaded from the botocore git submodule.

To test the new service, I've added a new integration test which performs a full CodeDeploy deployment to an on-premise instance. The instance is actually fake, and is simulated by the test itself by using the new rusoto_codedeploy_commands crate to process commands in place of the real agent.

Example test output from `cargo test --features codedeploy-commands -- --nocapture` is shown below.
```text
     Running target/debug/deps/codedeploy_commands-4beb864d44069378

running 1 test
Getting integration test identity
Creating on premise IAM role
Creating CodeDeploy service IAM role
Retrieving on-premise session credentials for simulated instance
Creating CodeDeploy application state
Registering simulated host as CodeDeploy on-premise instance
Starting deployment to simulated on-premise instance
Processing all deployment commands for simulated instance
Polled command ApplicationStop, going to acknowledge and complete it
Polled command DownloadBundle, going to acknowledge and complete it
Polled command BeforeInstall, going to acknowledge and complete it
Polled command Install, going to acknowledge and complete it
Polled command AfterInstall, going to acknowledge and complete it
Polled command ApplicationStart, going to acknowledge and complete it
Polled command ValidateService, going to acknowledge and complete it
Waiting for deployment to successfully complete
Deregistering simulated host from CodeDeploy
Deleting CodeDeploy application state
Deleting CodeDeploy service IAM role
Deleting on-premise IAM role
test successful_deployment_to_host ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

drake-amz commented:
Didn't see that exact failure in other PRs I looked at, but I did see that semaphoreci seems to always fail.

drake-amz commented:
For the travis-ci failure, all Mac builds worked and Linux builds for Rust 1.23.0, beta, and nightly worked. The only failure was for Rust stable, which for some reason took more than 49 minutes 29 seconds which appears to be the max timeout.

drake-amz commented:
Although none of the builds failed due to my new integration test failing, I am having some trouble with it locally all of a sudden (IAM propagation delay getting worse?) so I'm going to make some improvements to it. I'm planning on not deleting the IAM roles created by the test, plus adding clean up logic to delete all CodeDeploy applications, deployment groups, and on-premise instances.

srijs commented:
Hi, thanks for your PR!

I'm okay with the overall approach, however I think it might be cleaner if embedded vs. non-embedded would be indicated through a flag in `services.json`, rather than having to probe both sources. What do you think about that?

With regards to the integration tests, the general rule we're trying to follow is that integration tests should always clean up all the resources they created. Maybe there is a way to wait/poll until IAM state has been fully propagated?

As for the semaphore CI failures, I'm fixing these via https://github.com/rusoto/rusoto/pull/1038. Once that is merged, you should be able to re-base on master to pick up the fixes.

drake-amz commented:
> I'm okay with the overall approach, however I think it might be cleaner if embedded vs. non-embedded would be indicated through a flag in services.json, rather than having to probe both sources. What do you think about that?

This is how I actually first wrote it, but I found a couple things that made me change it to the implicit way it works at the moment. First, the implementation was very slightly nicer in that `ServiceDefinition::load` didn't need its signature changed to add an `is_embedded: bool` flag, and as a result I was able to reuse some code between `load` and `load_all`. Second, for most crates there's no extra work done because the resolution logic looks in botocore first, and only if its not found there does it look in the embedded directory.

That said, if you think having an explicit flag in the services list is more understandable I'd be happy to add it back in.

> With regards to the integration tests, the general rule we're trying to follow is that integration tests should always clean up all the resources they created. Maybe there is a way to wait/poll until IAM state has been fully propagated?

Yeah, I appreciate the benefits of leaving no traces. This situation is a bit tricky. The test currently creates four resources:
1. IAM role for a simulated instance. 
2. CodeDeploy application and deployment group.
3. CodeDeploy on-premise instance registration.
4. CodeDeploy IAM service role.

Resources 1, 2, and 3 are cleaned up. Additionally, because 2 and 3 contain random elements in their names, I've added cleanup logic so that any resources left over from failed tests will be cleaned up the next time the test runs.

The problem is with resource 4. There are two places the CodeDeploy service tries to assume the role: synchronously when we call CreateDeployment, and asynchronously when it goes to actually run the deployment. I have implemented retries on the CreateDeployment call so that we retry (up to 30 seconds) the call if it fails due to the role not being propagated yet. Unfortunately that still leaves a failure mode which is that CodeDeploy is able to assume the role in CreateDeployment, but is unable to do so in the asynchronous follow up due to (presumably) hitting a different STS service host that hasn't gotten the role propagated to it yet. This manifests in the deployment failing with an error message saying the role doesn't exist, resulting in the integration test failing.

The test code as it stands works around this by not deleting the service role. If you want all resources to be deleted, the trade off will be that the test will need to pause to wait for the IAM propagation (I saw delays of up to 30 seconds in some cases). Even if I do add this wait, there will still be a small chance of failure.

drake-amz commented:
Thinking about the propagation problem some more, I think a good approach that allows us to delete the role but be reasonably confident the test will not fail spuriously is to allow the test user's identity to assume the role in addition to the CodeDeploy service principal. Then after we create the role, call AssumeRole every second for 30 seconds, resetting the counter when the AssumeRole call fails. This will enforce that we have seen the successful ability to assume the role for 30 seconds past the latest failure.

drake-amz commented:
Travis build still failing due to "The job exceeded the maximum time limit for jobs, and has been terminated."

srijs commented:
@drake-amz We fixed the Travis build yesterday, you should be able to re-base on master again to fix it. Sorry for the inconvenience!

drake-amz commented:
Integration test has been updated so that it now deletes all resources created during the test run, including IAM roles. It also has logic for cleaning up any stale resources from previous test runs when the test starts, so that failed tests don't leave resources permanently cluttering up the test account.

Updated output from `cargo test --features codedeploy-commands -- --nocapture`:
```text
     Running target/debug/deps/codedeploy_commands-4beb864d44069378

running 1 test
Getting integration test identity
Creating on premise IAM role
Finding CodeDeploy's managed policy ARN
Deleting stale CodeDeploy service IAM roles
Creating CodeDeploy service IAM role
Waiting for CodeDeploy service role to be consistently assumable
Assume role failed: Unknown("<ErrorResponse xmlns=\"https://sts.amazonaws.com/doc/2011-06-15/\">\n  <Error>\n    <Type>Sender</Type>\n    <Code>AccessDenied</Code>\n    <Message>Not authorized to perform sts:AssumeRole</Message>\n  </Error>\n  <RequestId>d7dad350-44bd-11e8-97db-df0385ed9bd4</RequestId>\n</ErrorResponse>\n")
Assume role failed: Unknown("<ErrorResponse xmlns=\"https://sts.amazonaws.com/doc/2011-06-15/\">\n  <Error>\n    <Type>Sender</Type>\n    <Code>AccessDenied</Code>\n    <Message>Not authorized to perform sts:AssumeRole</Message>\n  </Error>\n  <RequestId>d87f76e9-44bd-11e8-97db-df0385ed9bd4</RequestId>\n</ErrorResponse>\n")
Assume role failed: Unknown("<ErrorResponse xmlns=\"https://sts.amazonaws.com/doc/2011-06-15/\">\n  <Error>\n    <Type>Sender</Type>\n    <Code>AccessDenied</Code>\n    <Message>Not authorized to perform sts:AssumeRole</Message>\n  </Error>\n  <RequestId>d923f4ee-44bd-11e8-97db-df0385ed9bd4</RequestId>\n</ErrorResponse>\n")
Assume role failed: Unknown("<ErrorResponse xmlns=\"https://sts.amazonaws.com/doc/2011-06-15/\">\n  <Error>\n    <Type>Sender</Type>\n    <Code>AccessDenied</Code>\n    <Message>Not authorized to perform sts:AssumeRole</Message>\n  </Error>\n  <RequestId>d9c8997c-44bd-11e8-97db-df0385ed9bd4</RequestId>\n</ErrorResponse>\n")
Assume role failed: Unknown("<ErrorResponse xmlns=\"https://sts.amazonaws.com/doc/2011-06-15/\">\n  <Error>\n    <Type>Sender</Type>\n    <Code>AccessDenied</Code>\n    <Message>Not authorized to perform sts:AssumeRole</Message>\n  </Error>\n  <RequestId>da6cc806-44bd-11e8-97db-df0385ed9bd4</RequestId>\n</ErrorResponse>\n")
Assume role failed: Unknown("<ErrorResponse xmlns=\"https://sts.amazonaws.com/doc/2011-06-15/\">\n  <Error>\n    <Type>Sender</Type>\n    <Code>AccessDenied</Code>\n    <Message>Not authorized to perform sts:AssumeRole</Message>\n  </Error>\n  <RequestId>db116d31-44bd-11e8-97db-df0385ed9bd4</RequestId>\n</ErrorResponse>\n")
Assume role failed: Unknown("<ErrorResponse xmlns=\"https://sts.amazonaws.com/doc/2011-06-15/\">\n  <Error>\n    <Type>Sender</Type>\n    <Code>AccessDenied</Code>\n    <Message>Not authorized to perform sts:AssumeRole</Message>\n  </Error>\n  <RequestId>dbb59c04-44bd-11e8-97db-df0385ed9bd4</RequestId>\n</ErrorResponse>\n")
Saw 30 consecutive AssumeRole successes after 36 total attempts
Retrieving on-premise session credentials for simulated instance
On-premise session ARN for simulated instance is arn:aws:sts::blahblahblah:assumed-role/rusoto-cd-commands-test-succ-dep/simulated-instance-1d646bd1-97b4-443f-9c83-04f03a85232c
Deleting stale CodeDeploy applications
Deregistering stale CodeDeploy on-premise instances
Creating CodeDeploy application state
Registering simulated host as CodeDeploy on-premise instance
Starting deployment to simulated on-premise instance
Deployment id is d-N562JJ7PS
Processing all deployment commands for simulated instance
Polling with host_identifier = arn:aws:sts::blahblahblah:assumed-role/rusoto-cd-commands-test-succ-dep/simulated-instance-1d646bd1-97b4-443f-9c83-04f03a85232c
Polled command ApplicationStop, going to acknowledge and complete it
Polling with host_identifier = arn:aws:sts::blahblahblah:assumed-role/rusoto-cd-commands-test-succ-dep/simulated-instance-1d646bd1-97b4-443f-9c83-04f03a85232c
Polled command DownloadBundle, going to acknowledge and complete it
Polling with host_identifier = arn:aws:sts::blahblahblah:assumed-role/rusoto-cd-commands-test-succ-dep/simulated-instance-1d646bd1-97b4-443f-9c83-04f03a85232c
Polled command BeforeInstall, going to acknowledge and complete it
Polling with host_identifier = arn:aws:sts::blahblahblah:assumed-role/rusoto-cd-commands-test-succ-dep/simulated-instance-1d646bd1-97b4-443f-9c83-04f03a85232c
Polled command Install, going to acknowledge and complete it
Polling with host_identifier = arn:aws:sts::blahblahblah:assumed-role/rusoto-cd-commands-test-succ-dep/simulated-instance-1d646bd1-97b4-443f-9c83-04f03a85232c
Polled command AfterInstall, going to acknowledge and complete it
Polling with host_identifier = arn:aws:sts::blahblahblah:assumed-role/rusoto-cd-commands-test-succ-dep/simulated-instance-1d646bd1-97b4-443f-9c83-04f03a85232c
Polled command ApplicationStart, going to acknowledge and complete it
Polling with host_identifier = arn:aws:sts::blahblahblah:assumed-role/rusoto-cd-commands-test-succ-dep/simulated-instance-1d646bd1-97b4-443f-9c83-04f03a85232c
Polled command ValidateService, going to acknowledge and complete it
Waiting for deployment to successfully complete
test successful_deployment_to_host ... test successful_deployment_to_host has been running for over 60 seconds
Deregistering simulated host from CodeDeploy
Deleting CodeDeploy application state
Deleting CodeDeploy service IAM role
Deleting on-premise IAM role
test successful_deployment_to_host ... ok
```

I have **not** added any flag to `services.json`, please see my remarks above and let me know if you'd still prefer I do so before merging.

matthewkmayer commented:
I plan on looking at this after merging in a couple new services and maybe the one to allow mocking of Rusoto futures. 👍

matthewkmayer commented:
@drake-amz I put some questions in https://github.com/rusoto/rusoto/issues/1023 . I'd love to hear more about this service and PR over there. ❤️

matthewkmayer commented:
Since this PR is stale and I don't have a good understanding of it or if it should go in Rusoto, I'll be closing this in the next week or so. 

If someone can answer the questions in the issue for this PR I'd love it! ❤️
